src/expr/src/scalar.rs

```diff
@@ -858,6 +858,12 @@ impl MirScalarExpr {
                             Some(vec![then, els])
                         }
                         MirScalarExpr::CallVariadic { func, exprs } => {
+                            // `Coalesce` conditionally evaluates its arguments based on their nullability.
+                            // The thinking is that the final term will be unconditionally evaluated, once
+                            // reached, which makes it suitable for optimization that relies on unstable
+                            // information like nullability. With further detail about the evaluation order
+                            // semantics, we may be able to simplify this logic, but until we are certain
+                            // we can avoid optimizing the terms based on an ephemeral nullability reading.
```

Comment:
> The thinking is that the final term will be unconditionally evaluated, once
> reached, 

Well, unless I'm misunderstand something, this is true for all arguments: once an argument is reached, it is unconditionally evaluated. Maybe you meant something like "the final argument can't prevent more arguments from being evaluated"?

Reply:
I think that is what I meant. I have to admit not having the right words to describe the bad thing that happens.